From 9cd73aa9fc2a10126643e5f8a00323cea35b2c72 Mon Sep 17 00:00:00 2001
From: Stephane Danieau <stephane.danieau@st.com>
Date: Wed, 7 Mar 2018 16:28:42 +0100
Subject: [PATCH 28/37] compositor-st: lastly call DRM_IOCTL_MODE_PAGE_FLIP in
 drm_output_repaint

In drm_output_repaint function, change the calling order from:
DRM_IOCTL_MODE_PAGE_FLIP
DRM_IOCTL_MODE_SETPLANE
DRM_IOCTL_WAIT_VBLANK
to
DRM_IOCTL_MODE_SETPLANE
DRM_IOCTL_WAIT_VBLANK
DRM_IOCTL_MODE_PAGE_FLIP

This fix the vblank handler that was not fired when requested due to
atomic feature alignment in the sti drm kernel driver.

Note: DO not upstream this patch. It is just a fix waiting for the
weston atomic functionality.

Change-Id: I928a365d546960a04a1125b4fda553e59767ca22
Signed-off-by: Vincent Abriou <vincent.abriou@st.com>
Reviewed-on: https://gerrit.st.com/63825
Reviewed-by: Fabien DESSENNE <fabien.dessenne@st.com>
Tested-by: Fabien DESSENNE <fabien.dessenne@st.com>
---
 libweston/compositor-st.c | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/libweston/compositor-st.c b/libweston/compositor-st.c
index 50473f8..2c1c829 100644
--- a/libweston/compositor-st.c
+++ b/libweston/compositor-st.c
@@ -889,16 +889,6 @@ drm_output_repaint(struct weston_output *output_base,
 		output_base->set_dpms(output_base, WESTON_DPMS_ON);
 	}
 
-	weston_xlog(" drmModePageFlip output->crtc_id = %d\n", output->crtc_id);
-	if (drmModePageFlip(backend->drm.fd, output->crtc_id,
-			    output->next->fb_id,
-			    DRM_MODE_PAGE_FLIP_EVENT, output) < 0) {
-		weston_log("queueing pageflip failed: %m\n");
-		goto err_pageflip;
-	}
-
-	output->page_flip_pending = 1;
-
 	drm_output_set_cursor(output);
 
 	/*
@@ -978,6 +968,16 @@ drm_output_repaint(struct weston_output *output_base,
 		output->vblank_pending++;
 	}
 
+	weston_xlog(" drmModePageFlip output->crtc_id = %d\n", output->crtc_id);
+	if (drmModePageFlip(backend->drm.fd, output->crtc_id,
+			    output->next->fb_id,
+			    DRM_MODE_PAGE_FLIP_EVENT, output) < 0) {
+		weston_log("queueing pageflip failed: %m\n");
+		goto err_pageflip;
+	}
+
+	output->page_flip_pending = 1;
+
 	return 0;
 
 err_pageflip:
-- 
1.9.1

